### ISSUE TITLE
Simple service server: no response is sent

### ISSUE THREAD

**Issue by anaelle-sw** [micro-ROS/micro_ros_arduino#9]
Hi! I am trying to create an "add_two_ints" service server, based on the example from micro-ROS-demos repo [here](https://github.com/micro-ROS/micro-ROS-demos/blob/foxy/rclc/addtwoints_server/main.c), but modified to get the " `void_setup()` / `void_loop()` format ". The server receives the request, but no response is sent to the service client.

## Setup:
- Computer: ROS2 Foxy / Ubuntu 20.04
- Board: Teensy 3.1 / Arduino IDE 1.8.13 / Teensyduino 1.53
Board is connected to computer via USB cable.

## Steps to reproduce:
The .ino source code uploaded on the Teensy is:
 ```
#include <micro_ros_arduino.h>
#include <example_interfaces/srv/add_two_ints.h>
#include <stdio.h>
#include <rcl/error_handling.h>
#include <rclc/rclc.h>
#include <rclc/executor.h>
#include <std_msgs/msg/int64.h>

// ros:
rcl_init_options_t options;
rcl_node_options_t node_ops;
rcl_node_t node;
rclc_support_t support;
rcl_allocator_t allocator;
rclc_executor_t executor;

// service:
rcl_service_options_t service_options;
rcl_service_t serv;
const rosidl_service_type_support_t * service_type_support;
rcl_wait_set_t wait_set;

// **** SETUP ****
void setup() {
  delay(5000); 
  allocator = rcl_get_default_allocator();

  options = rcl_get_zero_initialized_init_options();
  rcl_init_options_init(&options, rcl_get_default_allocator());

  rclc_support_init(&support, 0, NULL, &allocator);

  node_ops = rcl_node_get_default_options();
  node = rcl_get_zero_initialized_node();
  rclc_node_init_default(&node, "micro_ros_service_server_node", "", &support);

  // init server
  const char * service_name = "/add_two_ints";
  service_options = rcl_service_get_default_options();
  serv = rcl_get_zero_initialized_service();
  service_type_support = ROSIDL_GET_SRV_TYPE_SUPPORT(example_interfaces, srv, AddTwoInts);
  rcl_service_init(&serv, &node, service_type_support, service_name, &service_options);
  wait_set = rcl_get_zero_initialized_wait_set();
  rcl_wait_set_init(&wait_set, 0, 0, 0, 1, 0, 0, &support.context, rcl_get_default_allocator());

  executor = rclc_executor_get_zero_initialized_executor();
  rclc_executor_init(&executor, &support.context, 1, &allocator);
}


//**** LOOP ****
void loop() {
  delay(100);
  rcl_wait_set_clear(&wait_set);

  size_t index;
  rcl_wait_set_add_service(&wait_set, &serv, &index);

  rcl_wait(&wait_set, RCL_MS_TO_NS(1));
  for (size_t i = 0; i < wait_set.size_of_services; i++) {
    if (wait_set.services[i]) {

      rmw_request_id_t req_id;

      example_interfaces__srv__AddTwoInts_Request req;
      example_interfaces__srv__AddTwoInts_Request__init(&req);

      rcl_take_request(&serv, &req_id, &req);

      example_interfaces__srv__AddTwoInts_Response res;
      example_interfaces__srv__AddTwoInts_Response__init(&res);

      res.sum = req.a + req.b;

      rcl_send_response(&serv, &req_id, &res);
    }
  }
  rclc_executor_spin_some(&executor, RCL_MS_TO_NS(100));
}
```
Once the micro-ROS agent is running, I got the output:
```
[1604311603.418290] info     | TermiosAgentLinux.cpp | init                     | running...             | fd: 3
[1604311603.918485] info     | TermiosAgentLinux.cpp | fini                     | server stopped         | fd: 3
[1604311603.918687] info     | TermiosAgentLinux.cpp | init                     | running...             | fd: 3

[1604311607.854237] info     | Root.cpp           | create_client            | create                 | client_key: 0x01A06D5E, session_id: 0x81
[1604311607.854339] info     | SessionManager.hpp | establish_session        | session established    | client_key: 0x27290974, address: 1
[1604311607.854525] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 19, data: 
0000: 81 00 00 00 04 01 0B 00 00 00 58 52 43 45 01 00 01 0F 00
[1604311607.857389] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 116, data: 
0000: 81 80 00 00 01 05 6C 00 00 0A 00 01 01 02 00 00 5E 00 00 00 3C 64 64 73 3E 3C 70 61 72 74 69 63
0020: 69 70 61 6E 74 3E 3C 72 74 70 73 3E 3C 6E 61 6D 65 3E 6D 69 63 72 6F 5F 72 6F 73 5F 73 65 72 76
0040: 69 63 65 5F 73 65 72 76 65 72 5F 6E 6F 64 65 3C 2F 6E 61 6D 65 3E 3C 2F 72 74 70 73 3E 3C 2F 70
0060: 61 72 74 69 63 69 70 61 6E 74 3E 3C 2F 64 64 73 3E 00 00 00
[1604311607.863867] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0B 01 05 00 00 00 00 00 80
[1604311607.865046] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0B 01 05 00 00 00 00 00 80
[1604311607.865900] debug    | ProxyClient.cpp    | create_participant       | participant created    | client_key: 0x01A06D5E, participant_id: 0x000(1)
[1604311607.865946] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 14, data: 
0000: 81 80 00 00 05 01 06 00 00 0A 00 01 00 00
[1604311607.865958] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 01 00 00 00 80
[1604311607.865965] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 01 00 00 00 80
[1604311607.865972] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 01 00 00 00 80
[1604311607.867175] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0B 01 05 00 00 00 00 00 80
[1604311607.867186] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 01 00 00 00 80
[1604311607.867264] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 01 00 00 00 80
[1604311607.879828] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 352, data: 
0000: 81 80 01 00 01 05 58 01 00 0B 00 08 08 02 00 00 4A 01 00 00 3C 64 64 73 3E 3C 72 65 70 6C 69 65
0020: 72 20 70 72 6F 66 69 6C 65 5F 6E 61 6D 65 3D 22 30 5F 38 22 20 73 65 72 76 69 63 65 5F 6E 61 6D
0040: 65 3D 22 2F 61 64 64 5F 74 77 6F 5F 69 6E 74 73 22 20 72 65 71 75 65 73 74 5F 74 79 70 65 3D 22
0060: 65 78 61 6D 70 6C 65 5F 69 6E 74 65 72 66 61 63 65 73 3A 3A 73 72 76 3A 3A 64 64 73 5F 3A 3A 41
0080: 64 64 54 77 6F 49 6E 74 73 5F 52 65 71 75 65 73 74 5F 22 20 72 65 70 6C 79 5F 74 79 70 65 3D 22
00A0: 65 78 61 6D 70 6C 65 5F 69 6E 74 65 72 66 61 63 65 73 3A 3A 73 72 76 3A 3A 64 64 73 5F 3A 3A 41
00C0: 64 64 54 77 6F 49 6E 74 73 5F 52 65 73 70 6F 6E 73 65 5F 22 3E 3C 72 65 71 75 65 73 74 5F 74 6F
00E0: 70 69 63 5F 6E 61 6D 65 3E 72 71 2F 61 64 64 5F 74 77 6F 5F 69 6E 74 73 52 65 71 75 65 73 74 3C
0100: 2F 72 65 71 75 65 73 74 5F 74 6F 70 69 63 5F 6E 61 6D 65 3E 3C 72 65 70 6C 79 5F 74 6F 70 69 63
0120: 5F 6E 61 6D 65 3E 72 72 2F 61 64 64 5F 74 77 6F 5F 69 6E 74 73 52 65 70 6C 79 3C 2F 72 65 70 6C
0140: 79 5F 74 6F 70 69 63 5F 6E 61 6D 65 3E 3C 2F 72 65 70 6C 69 65 72 3E 3C 2F 64 64 73 3E 00 00 01
[1604311607.879877] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0B 01 05 00 01 00 01 00 80
[1604311607.879884] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0B 01 05 00 01 00 01 00 80
[1604311607.879892] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0B 01 05 00 01 00 01 00 80
[1604311607.880998] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0B 01 05 00 01 00 01 00 80
[1604311607.881208] debug    | ProxyClient.cpp    | create_replier           | replier created        | client_key: 0x01A06D5E, requester_id: 0x000(7), participant_id: 0x000(1)
[1604311607.881269] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 14, data: 
0000: 81 80 01 00 05 01 06 00 00 0B 00 08 00 00
[1604311607.881280] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 02 00 00 00 80
[1604311607.881287] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 02 00 00 00 80
[1604311607.881293] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 02 00 00 00 80
[1604311607.881301] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 02 00 00 00 80
[1604311607.881307] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 02 00 00 00 80
[1604311607.886692] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0B 01 05 00 01 00 01 00 80
[1604311607.886751] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 02 00 00 00 80
[1604311607.887799] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 02 00 00 00 80
[1604311607.987966] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 24, data: 
0000: 81 80 02 00 08 01 10 00 00 0C 00 08 80 00 00 01 FF FF 00 00 00 00 00 00
[1604311607.988130] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 03 00 00 00 80
[1604311608.087934] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0B 01 05 00 01 00 02 00 80
[1604311608.088091] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 03 00 00 00 80
[1604311608.089162] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0B 01 05 00 01 00 02 00 80
[1604311608.089334] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 03 00 00 00 80
[1604311608.188686] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0B 01 05 00 02 00 02 00 80
[1604311608.188825] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 03 00 00 00 80
[1604311608.390812] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0B 01 05 00 02 00 02 00 80
[1604311608.391055] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 03 00 00 00 80
```

Then I call the service in another shell with the command: 
`ros2 service call add_two_ints example_interfaces/srv/AddTwoInts "{a: 1, b: 0}"`
, which output is:
```
waiting for service to become available...
requester: making request: example_interfaces.srv.AddTwoInts_Request(a=1, b=0) 
```

The micro-ROS agent response is:
```
[1604311838.727954] debug    | Replier.cpp        | read_fn                  | [==>> DDS <<==]        | client_key: 0x00000000, len: 40, data: 
0000: 01 0F 8E E6 C1 71 01 00 01 00 00 00 00 00 12 03 00 00 00 00 01 00 00 00 01 00 00 00 00 00 00 00
0020: 00 00 00 00 00 00 00 00
[1604311838.728068] debug    | SerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x01A06D5E, len: 52, data: 
0000: 81 80 02 00 09 01 2C 00 00 0C 00 08 01 0F 8E E6 C1 71 01 00 01 00 00 00 00 00 12 03 00 00 00 00
0020: 01 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
[1604311838.745935] debug    | SerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x01A06D5E, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 03 00 00 00 80
```
So, the service is recognized as available, and the server receives the request, but the response is not sent. 

I also tried to launch the micro-ROS agent via a docker: 
` sudo docker run -it --rm -v /dev:/dev --privileged --net=host microros/micro-ros-agent:foxy serial --dev /dev/teensy -v6`
The outputs are the same.

I guess this issue is due to my .ino source code. Do you have any idea what causes this? Thanks for helping!

**Comment by pablogs9** [comment]
Working on it, I'll tell you something asap.

**Comment by pablogs9** [comment]
Hello @anaelle-sw I found something:

Please recheck [this line](https://github.com/micro-ROS/micro-ROS-demos/blob/e4cc345159f9b0dcb578a39020311922af0ec150/rclc/addtwoints_server/main.c#L31).

You have written: 
```c
 rcl_wait_set_init(&wait_set, 0, 0, 0, 1, 0, 0, &support.context, rcl_get_default_allocator());
```

and in the demo it says:

```c
rcl_wait_set_init(&wait_set, 0, 0, 0, 0, 1, 0, &context, rcl_get_default_allocator())
```

You are not creating a wait set with one service "slot". Check the function [here](https://github.com/micro-ROS/rcl/blob/20e2496ddc846cdda38d25c1c284b23e24735230/rcl/include/rcl/wait.h#L132).

Close if you are ok with this answer, I'm going to add to this repo a service example!

----

Other considerations, services, and clients are not fully supported in the RCLC (the micro-ROS user API), and you have to use the RCL. This is going to be fixed as soon as we can, but by now please take care when mixing both APIs. I suggest that you can remove the ` rclc_executor_spin_some(&executor, RCL_MS_TO_NS(100))` since you are actually doing a `rcl_wait()` previously.

Of course, you can mix the call to both, but please be aware of this.

**Comment by pablogs9** [comment]
You have my working example [here](https://github.com/micro-ROS/micro_ros_arduino/pull/12)

**Comment by anaelle-sw** [comment]
Hi! Thanks for these explanations, especially about the RCLC / RCL libraries. I was not really being careful with using one or the other.

I actually tried to use the function as used in the demo:
`rcl_wait_set_init(&wait_set, 0, 0, 0, 0, 1, 0, &context, rcl_get_default_allocator())`
This way, the service client did detect the service as available, but the micro-ROS agent wasn't not notifying that any request was received. So I thought I was closer to a solution using `&support.context`, since the request was received.
Anyway, your example server is working nicely for me, thanks!

Just a quick question before closing this issue please: in your example service server code, you initialize an executor in the `void setup()` function [here](https://github.com/micro-ROS/micro_ros_arduino/blob/1b00ac36393e849fb4c7f6c658c76d433527208f/examples/micro-ros_addtwoints_service/micro-ros_addtwoints_service.ino#L48), and then in the `void_loop()` function you use a `rcl_wait()` [here](https://github.com/micro-ROS/micro_ros_arduino/blob/1b00ac36393e849fb4c7f6c658c76d433527208f/examples/micro-ros_addtwoints_service/micro-ros_addtwoints_service.ino#L60). As you explained, it is not necessary to use `rclc_executor_spin_some()`. So why you actually need to initialize an executor anyway? Sorry if this question does not belong here, it may be quite related to "general" micro-ROS

**Comment by pablogs9** [comment]
Hello @anaelle-sw,

You are right, if in the loop we are using a `rcl_waitset` directly, the executor is not needed. In fact, as you can see [here](https://github.com/micro-ROS/rclc/tree/master/rclc), the RCLC executor is just a sort of intelligent wrapper for the `waitset`, internally the executor handles its own waitset and allows the user setting callbacks, managing the execution order and so on...

So basically, we are using here the rcl_waitset because we don't have yet support for services in the executor. As soon as we merge [this](https://github.com/micro-ROS/rclc/pull/44), you can forget the waitsets and work with the RCLC API and its callbacks.

I hope that this is not a big inconvenient.

**Comment by anaelle-sw** [comment]
Ok, I understand better now.
Actually, I am as new to RCL than to RCLC, so it won't be a big deal in my case. 
I will close this issue. Thanks for your time!

**Comment by pablogs9** [comment]
Related: Check [here](https://github.com/micro-ROS/micro_ros_arduino/blob/foxy/examples/micro-ros_addtwoints_service/micro-ros_addtwoints_service.ino) the new service/client API in RCLC.

**Comment by anaelle-sw** [comment]
Very nice to have service callbacks directly in the API. This way, it is really lighter and more intuitive to use. Thanks for notifying me about it!

**Comment by anaelle-sw** [comment]
Just three quick questions about this new RCLC API and your example code:
_(if you want, I can open a new issue instead of posting here, let me know)_

1) In your example code, you declare a `rcl_wait_set_t` but don't initialize it, neither use it (as well as a `rmw_request_id_t` and a `rosidl_service_type_support_t` by the way). In my case, it works well if these are not declared. If I don't misunderstand, a wait set is not required anymore with the RCLC API, as you have an executor spinning (`rclc_executor_spin_some` in the loop), and RCLC directly handles this, is this right?

2) I don't need the `res` and `req` outside the service callback. Can I just don't declare these are global variables, and go with
```
rclc_executor_add_service(
                       &executor,
                       &service, 
                       &example_interfaces__srv__AddTwoInts_Request(), 
                       &example_interfaces__srv__AddTwoInts_Response(),
                       service_callback);
```
when initializing my service? My code builds this way, and everything seems to work normally. But can this lead to known problems? (I am thinking about memory leak for instance)

3) Is there a way to wait for the response to be sent, inside the callback? Actually, I need to copy strings inside `res->message`. As you told me in [this post](https://github.com/micro-ROS/micro_ros_arduino/issues/23#issuecomment-724497240), I cannot directly copy paste strings, but must use memory allocation and memory copy. For instance, when following this instruction, one of my services' callbacks looks like this:
```
void trigger_srv_cb(const void * req_in, void * res_in){
  std_srvs__srv__Trigger_Request * req = (std_srvs__srv__Trigger_Request *) req_in;
  std_srvs__srv__Trigger_Response * res = (std_srvs__srv__Trigger_Response *) res_in;
 
  res->message.data = (char*)malloc(200 * sizeof(char));
  res->message.size = 0;
  res->message.capacity = 200;
  memcpy(res->message.data, "trigger executed", 200);
  res->message.size = strlen(res->message.data);

  free(res->message.data);
}
```
The problem is I need to execute `free` after `malloc` to avoid memory leak. With the previous API, I call the function `rcl_send_response` before doing `free`. But with RCLC API (which is really more convenient, and that I am really willing to use) the response might not be sent before executing `free`. What is the correct way to handle this?

**Comment by pablogs9** [comment]
Hello @anaelle-sw just I forgot to merge https://github.com/micro-ROS/micro_ros_arduino/pull/45, so:
1. solved in https://github.com/micro-ROS/micro_ros_arduino/pull/45

2. Wow, this is tricky. Where do you have defined `example_interfaces__srv__AddTwoInts_Response()` I'm not sure if they are pure C99 implementation or  Arduino C++. Usually you have to provide a req and res object to the executor to let it have some memory to save the req and res since the executor is not allowed to alloc memory. In Arduino I usually put all the variables as global since `setup()` stack variables can't be used in `loop()`... Does your solution work?

3. If you follow the example we provide, you have `req` and `res` declared as global variables. So I suggest you to do the 
```c
example_interfaces__srv__AddTwoInts_Response res;
example_interfaces__srv__AddTwoInts_Request req;

void setup(){
...
  res.message.data = (char*)malloc(200 * sizeof(char));
  res.message.size = 0;
  res.message.capacity = 200;
  RCCHECK(rclc_executor_add_service(&executor, &service, &req, &res, service_callback));
....
```
in the `setup()`. Then you provide those two object to the executor and you have guaranteed that in the callback the executor will pass you the couple of pointers that are the same that you provide: so the res is already init and allocated with the malloc. So in the callback you just do:
```c
void trigger_srv_cb(const void * req_in, void * res_in){
  std_srvs__srv__Trigger_Request * req = (std_srvs__srv__Trigger_Request *) req_in;
  std_srvs__srv__Trigger_Response * res = (std_srvs__srv__Trigger_Response *) res_in;
 
  memcpy(res->message.data, "trigger executed", 200);   <-- be careful with this 200, your string is not 200!
  res->message.size = strlen(res->message.data);
}
```

**Also be SUPER careful with copying not NULL-ending string because serialization for strings relies in a '\0' at the end... Maybe is safer a `fprintf(res->message.data, "%s", "blablabla")`.**

**Comment by anaelle-sw** [comment]
2) Well, I didn't defined `example_interfaces__srv__AddTwoInts_Response()` actually, just included the file `example_interfaces/srv/add_two_ints.h`. Yes the code works this way, but I just tested it with your example code for now, with a few service calls.

3) Ok then, I will follow your advice and just use global variables (I am used to limit the number of global variables when programming ROS nodes for instance, but in the case of Arduino code I guess it is not a problem).

The problem with `fprintf(res->message.data, "%s", "blablabla")` is that you can only see this debug string via the serial monitor provided by Arduino IDE, is this right? But here we would need to get the `res->message` in ROS, or at least in a terminal. 

> ```c
>  be careful with this 200, your string is not 200!
> ```

This could be a quite big problem with our application. We need to regularly copy strings in a message (which is published by our board). The problem is the strings have different length. So having something like this:
 `memcpy(res->message.data, "trigger executed", sizeof("trigger executed"));`
, and later on something like this (on the same `res->message`): 
`memcpy(res->message.data, "foo", sizeof("foo"));`
, results on publishing a message like this: 
`"foogger executed"`.

My solution for this was to set a maximum string length(` #define MAX_LEN_STR 100`), and use it every time I need to copy a string:  
`memcpy(res->message.data, "trigger executed", MAX_LEN_STR);`. 
For now, it works very well with our string-containing message publication, and a ROS node processing the strings. The tests I ran for now lasted around 20min max, no problem related to the published strings occurred. What can happen if the size given in `memcpy` is not the one of the copied string?

**Comment by pablogs9** [comment]
Sorry I didn't write the function well. I didn't mean `fprintf` I meant `sprintf`.

You are having the "foogger executed" string because you are not handling the '\0' character for null-ending the string. My solution for copying a string to a buffer is:

```c
sprintf(res->message.data, "%s", "blablabla")
res->message.size = strlen(res->message.data);
```

This writes "blablabla\0" to your buffer, so strlen will work to obtain the size.

> What can happen if the size given in memcpy is not the one of the copied string?

Multiple things:
- You are copying unnecessary bytes to a buffer, and at some point, this is inefficient
- Imagine that you are telling the code to copy `"blabla"` to `buf[100]` with `memcpy(buf, "blabla", 100);`. You know that `buf` has 100 slots to write but does the literal `"blabla"` (which is in the flash .const section) has 100 characters for reading? Imagine your string is stored at the end of the memory, in this case you will do an illegal read. Imagine you are in Linux, probably you will have a segfault if you read from a section of memory that does not belong to your process.
- Your string are only NULL-terminated if the buffer is previously set to 0

**Comment by anaelle-sw** [comment]
Ok, I got it now. Thank you very much for all explanations!
